Fix integration tests with random failures

<!-- Please only use this template for submitting reports about failing tests in KubeEdge CI jobs -->

**Which jobs are failing**:
integration-test

**Which test(s) are failing**:
device tests

**Since when has it been failing**:
Random failures

**Reason for failure**:
device isn't in the context before publish messages to change the state are posted.

**Anything else we need to know**:
Log : 
```sh
------------------------------
• Failure [60.002 seconds]
Event Bus Testing
/home/travis/gopath/src/github.com/kubeedge/kubeedge/edge/test/integration/device/device_test.go:36
  Publish on eventbus topics throgh MQTT internal broker
  /home/travis/gopath/src/github.com/kubeedge/kubeedge/edge/test/integration/device/device_test.go:40
    TC_TEST_EBUS_6: change the device status to online from eventbus [It]
    /home/travis/gopath/src/github.com/kubeedge/kubeedge/edge/test/integration/device/device_test.go:103
    Timed out after 60.000s.
    Device state is not online within specified time
    Expected
        <string>: unknown
    to equal
        <string>: online
    /home/travis/gopath/src/github.com/kubeedge/kubeedge/edge/test/integration/device/device_test.go:121
------------------------------
I0313 01:16:08.267387   11721 log.go:70] client.Publish Success !!
I0313 01:16:08.268051   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:08.268186   11721 log.go:70] TestCase:TC_TEST_EBUS_7: change the device status to unknown from eventbus     Status=PASSED
•I0313 01:16:08.269487   11721 log.go:70] client.Publish Success !!
I0313 01:16:08.288640   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:10.289727   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:12.290537   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:14.291630   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:16.292584   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:18.293373   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:20.294272   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:22.295125   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:24.296841   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:26.297880   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:28.298822   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:30.299599   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:32.300341   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:34.301235   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:36.302201   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:38.303965   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:40.304904   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:42.305817   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:44.306685   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:46.307537   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:48.308423   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:50.309306   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:52.310121   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:54.310985   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:56.311882   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:16:58.312780   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:17:00.313650   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:17:02.314514   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:17:04.316847   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:17:06.317866   11721 log.go:65] DeviceID= kubeedge-device-8tmyeg6gdp, DeviceState= unknown
I0313 01:17:08.270440   11721 log.go:70] TestCase:TC_TEST_EBUS_8: change the device status to offline from eventbus     Status=FAILED
------------------------------
• Failure [60.002 seconds]
Event Bus Testing
/home/travis/gopath/src/github.com/kubeedge/kubeedge/edge/test/integration/device/device_test.go:36
  Publish on eventbus topics throgh MQTT internal broker
  /home/travis/gopath/src/github.com/kubeedge/kubeedge/edge/test/integration/device/device_test.go:40
    TC_TEST_EBUS_8: change the device status to offline from eventbus [It]
    /home/travis/gopath/src/github.com/kubeedge/kubeedge/edge/test/integration/device/device_test.go:147
    Timed out after 60.000s.
    Device state is not offline within specified time
    Expected
        <string>: unknown
    to equal
        <string>: offline
    /home/travis/gopath/src/github.com/kubeedge/kubeedge/edge/test/integration/device/device_test.go:164
------------------------------
I0313 01:17:08.270569   11721 log.go:65] Remove Mock device from edgenode !!
I0313 01:17:08.272272   11721 log.go:70] DELETE http://127.0.0.1:12345/devices 200 OK in 1.64897ms
Summarizing 2 Failures:
[Fail] Event Bus Testing Publish on eventbus topics throgh MQTT internal broker [It] TC_TEST_EBUS_6: change the device status to online from eventbus 
/home/travis/gopath/src/github.com/kubeedge/kubeedge/edge/test/integration/device/device_test.go:121
[Fail] Event Bus Testing Publish on eventbus topics throgh MQTT internal broker [It] TC_TEST_EBUS_8: change the device status to offline from eventbus 
/home/travis/gopath/src/github.com/kubeedge/kubeedge/edge/test/integration/device/device_test.go:164
Ran 8 of 8 Specs in 120.018 seconds
FAIL! -- 6 Passed | 2 Failed | 0 Pending | 0 Skipped
--- FAIL: TestEdgecoreEventBus (120.02s)
FAIL
```
Failed job URL : https://travis-ci.org/kubeedge/kubeedge/jobs/505530549

--- a/edge/test/integration/device/device_test.go
+++ b/edge/test/integration/device/device_test.go
@@ -19,6 +19,7 @@ package device_test
 import (
 	"encoding/json"
 	"net/http"
+	"strings"
 
 	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
 	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dttype"
@@ -31,7 +32,60 @@ import (
 	. "github.com/onsi/gomega"
 )
 
-//getting deviceid from the DB and assigning to it
+//Devicestate from subscribed MQTT topic
+var DeviceState string
+
+type DeviceUpdates struct {
+	EventId     string `json:"event_id"`
+	Timestamp   int    `json:"timestamp"`
+	DeviceField `json:"device"`
+}
+
+type DeviceField struct {
+	Name       string `json:"name"`
+	State      string `json:"state"`
+	LastOnline string `json:"last_online"`
+}
+
+type MembershipUpdate struct {
+	BaseMessage
+	AddDevices    []Device `json:"added_devices"`
+	RemoveDevices []Device `json:"removed_devices"`
+}
+
+type BaseMessage struct {
+	EventID   string `json:"event_id"`
+	Timestamp int64  `json:"timestamp"`
+}
+
+var MemDeviceUpdate MembershipUpdate
+
+func SubMessageReceived(client MQTT.Client, message MQTT.Message) {
+	var deviceState DeviceUpdates
+	topic := dtcommon.DeviceETPrefix + DeviceIDN + dtcommon.DeviceETStateUpdateSuffix + "/result"
+	if message.Topic() == topic {
+		devicePayload := (message.Payload())
+		err := json.Unmarshal(devicePayload, &deviceState)
+		if err != nil {
+			common.Failf("Unmarshall failed %s", err)
+		}
+	}
+	DeviceState = deviceState.State
+	common.InfoV2("device updated %+v", deviceState)
+}
+func DeviceSubscribed(client MQTT.Client, message MQTT.Message) {
+	topic := dtcommon.MemETPrefix + ctx.Cfg.NodeId + dtcommon.MemETUpdateSuffix
+	if message.Topic() == topic {
+		devicePayload := (message.Payload())
+		err := json.Unmarshal(devicePayload, &MemDeviceUpdate)
+		if err != nil {
+			common.Failf("Unmarshall failed %s", err)
+		}
+	}
+	common.InfoV2("device list is %+v", MemDeviceUpdate)
+}
+
+// Deviceid from the DB and assigning to it
 var DeviceIDN string
 var DeviceN dttype.Device
 var DeviceIDWithAttr string
@@ -41,134 +95,161 @@ var DeviceTW dttype.Device
 
 //Run Test cases
 var _ = Describe("Event Bus Testing", func() {
-	var Token_client Token
+	var TokenClient Token
 	var ClientOpts *MQTT.ClientOptions
 	var Client MQTT.Client
 	Context("Publish on eventbus topics throgh MQTT internal broker", func() {
 		BeforeEach(func() {
 			ClientOpts = HubClientInit(ctx.Cfg.MqttEndpoint, ClientID, "", "")
 			Client = MQTT.NewClient(ClientOpts)
-			if Token_client = Client.Connect(); Token_client.Wait() && Token_client.Error() != nil {
-				common.Failf("client.Connect() Error is %s", Token_client.Error())
+			if TokenClient = Client.Connect(); TokenClient.Wait() && TokenClient.Error() != nil {
+				common.Failf("client.Connect() Error is %s", TokenClient.Error())
 			}
-			Expect(Token_client.Error()).NotTo(HaveOccurred())
+			Expect(TokenClient.Error()).NotTo(HaveOccurred())
+			devicetopic := dtcommon.MemETPrefix + ctx.Cfg.NodeId + dtcommon.MemETUpdateSuffix
+			topic := dtcommon.DeviceETPrefix + DeviceIDN + dtcommon.DeviceETStateUpdateSuffix + "/result"
+			Client.Subscribe(devicetopic, 0, DeviceSubscribed)
+			Client.Subscribe(topic, 0, SubMessageReceived)
 		})
 		AfterEach(func() {
 			common.PrintTestcaseNameandStatus()
 		})
 
 		It("TC_TEST_EBUS_1: Sending data to Cloud", func() {
-			if Token_client = Client.Publish(UploadRecordToCloud, 0, false, "messagetoUpload_record_to_cloud"); Token_client.Wait() && Token_client.Error() != nil {
-				common.Failf("client.Publish() Error is %s", Token_client.Error())
+
+			if TokenClient = Client.Publish(UploadRecordToCloud, 0, false, "messagetoUpload_record_to_cloud"); TokenClient.Wait() && TokenClient.Error() != nil {
+				common.Failf("client.Publish() Error is %s", TokenClient.Error())
 			} else {
 				common.InfoV6("client.Publish Success !!")
 			}
-			Expect(Token_client.Error()).NotTo(HaveOccurred())
+			Expect(TokenClient.Error()).NotTo(HaveOccurred())
 			Client.Disconnect(1)
 		})
 
 		It("TC_TEST_EBUS_2: Sending data to device module", func() {
-			if Token_client = Client.Publish(DevicestatusUpdate, 0, false, "messagetoDevice_status_update"); Token_client.Wait() && Token_client.Error() != nil {
-				common.Failf("client.Publish() Error is %s", Token_client.Error())
+			if TokenClient = Client.Publish(DevicestatusUpdate, 0, false, "messagetoDevice_status_update"); TokenClient.Wait() && TokenClient.Error() != nil {
+				common.Failf("client.Publish() Error is %s", TokenClient.Error())
 			} else {
 				common.InfoV6("client.Publish Success !!")
 			}
-			Expect(Token_client.Error()).NotTo(HaveOccurred())
+			Expect(TokenClient.Error()).NotTo(HaveOccurred())
 			Client.Disconnect(1)
 		})
 
 		It("TC_TEST_EBUS_3: Sending data to device twin module", func() {
-			if Token_client = Client.Publish(DeviceTwinUpdate, 0, false, "messagetoDevice_Twin_update"); Token_client.Wait() && Token_client.Error() != nil {
-				common.Failf("client.Publish() Error is %s", Token_client.Error())
+			if TokenClient = Client.Publish(DeviceTwinUpdate, 0, false, "messagetoDevice_Twin_update"); TokenClient.Wait() && TokenClient.Error() != nil {
+				common.Failf("client.Publish() Error is %s", TokenClient.Error())
 			} else {
 				common.InfoV6("client.Publish Success !!")
 			}
-			Expect(Token_client.Error()).NotTo(HaveOccurred())
+			Expect(TokenClient.Error()).NotTo(HaveOccurred())
 			Client.Disconnect(1)
 		})
 
 		It("TC_TEST_EBUS_4: Sending data to membership module", func() {
-			if Token_client = Client.Publish(DeviceMembershipUpdate, 0, false, "messagetoDevice_Membership_update"); Token_client.Wait() && Token_client.Error() != nil {
-				common.Failf("client.Publish() Error is %s", Token_client.Error())
+			if TokenClient = Client.Publish(DeviceMembershipUpdate, 0, false, "messagetoDevice_Membership_update"); TokenClient.Wait() && TokenClient.Error() != nil {
+				common.Failf("client.Publish() Error is %s", TokenClient.Error())
 			} else {
 				common.InfoV6("client.Publish Success !!")
 			}
-			Expect(Token_client.Error()).NotTo(HaveOccurred())
+			Expect(TokenClient.Error()).NotTo(HaveOccurred())
 			Client.Disconnect(1)
 		})
 
 		It("TC_TEST_EBUS_5: Sending data to device module", func() {
-			if Token_client = Client.Publish(DeviceUpload, 0, false, "messagetoDevice_upload"); Token_client.Wait() && Token_client.Error() != nil {
-				common.Failf("client.Publish() Error is %s", Token_client.Error())
+			if TokenClient = Client.Publish(DeviceUpload, 0, false, "messagetoDevice_upload"); TokenClient.Wait() && TokenClient.Error() != nil {
+				common.Failf("client.Publish() Error is %s", TokenClient.Error())
 			} else {
 				common.InfoV6("client.Publish Success !!")
 			}
-			Expect(Token_client.Error()).NotTo(HaveOccurred())
+			Expect(TokenClient.Error()).NotTo(HaveOccurred())
 			Client.Disconnect(1)
 		})
 
 		It("TC_TEST_EBUS_6: change the device status to online from eventbus", func() {
 			var message DeviceUpdate
 			message.State = "online"
+			topic := dtcommon.DeviceETPrefix + DeviceIDN + dtcommon.DeviceETStateUpdateSuffix + "/result"
 			body, err := json.Marshal(message)
 			if err != nil {
 				common.Failf("Marshal failed %v", err)
 			}
-
-			if Token_client = Client.Publish(dtcommon.DeviceETPrefix+DeviceIDN+dtcommon.DeviceETStateUpdateSuffix, 0, false, body); Token_client.Wait() && Token_client.Error() != nil {
-				common.Failf("client.Publish() Error is %s", Token_client.Error())
-			} else {
-				common.InfoV6("client.Publish Success !!")
-			}
-			Expect(Token_client.Error()).NotTo(HaveOccurred())
 			Eventually(func() string {
-				deviceState := GetDeviceStateFromDB(DeviceIDN)
-				common.InfoV2("DeviceID= %s, DeviceState= %s", DeviceIDN, deviceState)
-				return deviceState
-			}, "60s", "2s").Should(Equal("online"), "Device state is not online within specified time")
-
+				var deviceEvent Device
+				for _, deviceEvent = range MemDeviceUpdate.AddDevices {
+					if strings.Compare(deviceEvent.ID, DeviceIDN) == 0 {
+						if TokenClient = Client.Publish(dtcommon.DeviceETPrefix+DeviceIDN+dtcommon.DeviceETStateUpdateSuffix, 0, false, body); TokenClient.Wait() && TokenClient.Error() != nil {
+							common.Failf("client.Publish() Error is %s", TokenClient.Error())
+						} else {
+							common.InfoV6("client.Publish Success !!")
+						}
+					}
+				}
+				return deviceEvent.ID
+			}, "10s", "2s").Should(Equal(DeviceIDN), "Device state is not online within specified time")
+			Expect(TokenClient.Error()).NotTo(HaveOccurred())
+			Eventually(func() string {
+				common.InfoV2("subscribed to the topic %v", topic)
+				return DeviceState
+			}, "10s", "2s").Should(Equal("online"), "Device state is not online within specified time")
 			Client.Disconnect(1)
 		})
 
 		It("TC_TEST_EBUS_7: change the device status to unknown from eventbus", func() {
 			var message DeviceUpdate
 			message.State = "unknown"
+			topic := dtcommon.DeviceETPrefix + DeviceIDN + dtcommon.DeviceETStateUpdateSuffix + "/result"
 			body, err := json.Marshal(message)
 			if err != nil {
 				common.Failf("Marshal failed %v", err)
 			}
-			if Token_client = Client.Publish(dtcommon.DeviceETPrefix+DeviceIDN+dtcommon.DeviceETStateUpdateSuffix, 0, false, body); Token_client.Wait() && Token_client.Error() != nil {
-				common.Failf("client.Publish() Error is %s", Token_client.Error())
-			} else {
-				common.InfoV6("client.Publish Success !!")
-			}
-			Expect(Token_client.Error()).NotTo(HaveOccurred())
 			Eventually(func() string {
-				deviceState := GetDeviceStateFromDB(DeviceIDN)
-				common.InfoV2("DeviceID= %s, DeviceState= %s", DeviceIDN, deviceState)
-				return deviceState
-			}, "60s", "2s").Should(Equal("unknown"), "Device state is not unknown within specified time")
+				var deviceEvent Device
+				for _, deviceEvent = range MemDeviceUpdate.AddDevices {
+					if strings.Compare(deviceEvent.ID, DeviceIDN) == 0 {
+						if TokenClient = Client.Publish(dtcommon.DeviceETPrefix+DeviceIDN+dtcommon.DeviceETStateUpdateSuffix, 0, false, body); TokenClient.Wait() && TokenClient.Error() != nil {
+							common.Failf("client.Publish() Error is %s", TokenClient.Error())
+						} else {
+							common.InfoV6("client.Publish Success !!")
+						}
+					}
+				}
+				return deviceEvent.ID
+			}, "10s", "2s").Should(Equal(DeviceIDN), "Device state is not online within specified time")
+			Expect(TokenClient.Error()).NotTo(HaveOccurred())
+			Eventually(func() string {
+				common.InfoV2("subscribed to the topic %v", topic)
+				return DeviceState
+			}, "10s", "2s").Should(Equal("unknown"), "Device state is not unknown within specified time")
 			Client.Disconnect(1)
 		})
 
 		It("TC_TEST_EBUS_8: change the device status to offline from eventbus", func() {
 			var message DeviceUpdate
 			message.State = "offline"
+			topic := dtcommon.DeviceETPrefix + DeviceIDN + dtcommon.DeviceETStateUpdateSuffix + "/result"
 			body, err := json.Marshal(message)
 			if err != nil {
 				common.Failf("Marshal failed %v", err)
 			}
-			if Token_client = Client.Publish(dtcommon.DeviceETPrefix+DeviceIDN+dtcommon.DeviceETStateUpdateSuffix, 0, false, body); Token_client.Wait() && Token_client.Error() != nil {
-				common.Failf("client.Publish() Error is %s", Token_client.Error())
-			} else {
-				common.InfoV6("client.Publish Success !!")
-			}
-			Expect(Token_client.Error()).NotTo(HaveOccurred())
 			Eventually(func() string {
-				deviceState := GetDeviceStateFromDB(DeviceIDN)
-				common.InfoV2("DeviceID= %s, DeviceState= %s", DeviceIDN, deviceState)
-				return deviceState
-			}, "60s", "2s").Should(Equal("offline"), "Device state is not offline within specified time")
+				var deviceEvent Device
+				for _, deviceEvent = range MemDeviceUpdate.AddDevices {
+					if strings.Compare(deviceEvent.ID, DeviceIDN) == 0 {
+						if TokenClient = Client.Publish(dtcommon.DeviceETPrefix+DeviceIDN+dtcommon.DeviceETStateUpdateSuffix, 0, false, body); TokenClient.Wait() && TokenClient.Error() != nil {
+							common.Failf("client.Publish() Error is %s", TokenClient.Error())
+						} else {
+							common.InfoV6("client.Publish Success !!")
+						}
+					}
+				}
+				return deviceEvent.ID
+			}, "10s", "2s").Should(Equal(DeviceIDN), "Device state is not online within specified time")
+			Expect(TokenClient.Error()).NotTo(HaveOccurred())
+			Eventually(func() string {
+				common.InfoV2("subscribed to the topic %v", topic)
+				return DeviceState
+			}, "10s", "2s").Should(Equal("offline"), "Device state is not offline within specified time")
 			Client.Disconnect(1)
 		})
 

--- a/edge/test/integration/scripts/fast_test
+++ b/edge/test/integration/scripts/fast_test
@@ -31,7 +31,8 @@ cat >config.json<<END
         "mqttEndpoint":"tcp://$MQTT_SERVER:1884",
         "testManager": "http://127.0.0.1:12345",
         "edgedEndpoint": "http://127.0.0.1:10255",
-        "image_url": ["nginx:latest", "redis:latest"]
+        "image_url": ["nginx:latest", "redis:latest"],
+        "nodeId": "fb4ebb70-2783-42b8-b3ef-63e2fd6d242e"
 }
 END
 

--- a/edge/test/integration/utils/edge/config.go
+++ b/edge/test/integration/utils/edge/config.go
@@ -32,6 +32,7 @@ type Config struct {
 	TestManager   string   `json:"testManager"`
 	EdgedEndpoint string   `json:"edgedEndpoint"`
 	AppImageUrl   []string `json:"image_url"`
+	NodeId        string   `json:"nodeId"`
 }
 
 //config struct

--- a/edge/test/integration/utils/helpers/helpers.go
+++ b/edge/test/integration/utils/helpers/helpers.go
@@ -160,7 +160,6 @@ func GetDeviceStateFromDB(deviceID string) string {
 
 func GetTwinAttributesFromDB(deviceID string, Name string) TwinAttribute {
 	var twinAttribute TwinAttribute
-
 	pwd, err := os.Getwd()
 	if err != nil {
 		common.Failf("Failed to get PWD: %v", err)
@@ -197,7 +196,6 @@ func GetTwinAttributesFromDB(deviceID string, Name string) TwinAttribute {
 			break
 		}
 	}
-
 	return twinAttribute
 }
 

